File: app/src/main/java/com/example/englishhindi/repository/WordRepository.java
package com.bhashasetu.app.repository;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.bhashasetu.app.database.WordDao;
import com.bhashasetu.app.database.WordDatabase;
import com.bhashasetu.app.model.Word;

import java.util.List;

public class WordRepository {
    
    private WordDao wordDao;
    private LiveData<List<Word>> allWords;
    private LiveData<List<Word>> favoriteWords;
    
    public WordRepository(Application application) {
        WordDatabase database = WordDatabase.getInstance(application);
        wordDao = database.wordDao();
        allWords = wordDao.getAllWords();
        favoriteWords = wordDao.getFavoriteWords();
    }
    
    // Get the total number of words
    public int getWordCount() {
        return wordDao.getWordCount();
    }
    
    // Insert a new word
    public void insert(Word word) {
        new InsertWordAsyncTask(wordDao).execute(word);
    }
    
    // Update an existing word
    public void update(Word word) {
        new UpdateWordAsyncTask(wordDao).execute(word);
    }
    
    // Delete a word
    public void delete(Word word) {
        new DeleteWordAsyncTask(wordDao).execute(word);
    }
    
    // Delete all words
    public void deleteAllWords() {
        new DeleteAllWordsAsyncTask(wordDao).execute();
    }
    
    // Get all words
    public LiveData<List<Word>> getAllWords() {
        return allWords;
    }
    
    // Get words by category
    public LiveData<List<Word>> getWordsByCategory(String category) {
        return wordDao.getWordsByCategory(category);
    }
    
    // Get favorite words
    public LiveData<List<Word>> getFavoriteWords() {
        return favoriteWords;
    }
    
    // Search words
    public LiveData<List<Word>> searchWords(String query) {
        return wordDao.searchWords(query);
    }
    
    // Get least mastered words
    public LiveData<List<Word>> getLeastMasteredWords(int limit) {
        return wordDao.getLeastMasteredWords(limit);
    }
    
    // Get words by IDs
    public LiveData<List<Word>> getWordsByIds(List<Integer> wordIds) {
        return wordDao.getWordsByIds(wordIds);
    }
    
    // Get random words by difficulty
    public LiveData<List<Word>> getRandomWordsByDifficulty(int maxDifficulty, int limit) {
        return wordDao.getRandomWordsByDifficulty(maxDifficulty, limit);
    }
    
    // Get a specific word by ID
    public LiveData<Word> getWordById(int wordId) {
        return wordDao.getWordById(wordId);
    }
    
    // Get all categories
    public LiveData<List<String>> getAllCategories() {
        return wordDao.getAllCategories();
    }
    
    // Get words for a specific lesson
    public LiveData<List<Word>> getWordsForLesson(int lessonId) {
        return wordDao.getWordsForLesson(lessonId);
    }
    
    // AsyncTask classes for database operations
    private static class InsertWordAsyncTask extends AsyncTask<Word, Void, Void> {
        private WordDao wordDao;
        
        private InsertWordAsyncTask(WordDao wordDao) {
            this.wordDao = wordDao;
        }
        
        @Override
        protected Void doInBackground(Word... words) {
            wordDao.insert(words[0]);
            return null;
        }
    }
    
    private static class UpdateWordAsyncTask extends AsyncTask<Word, Void, Void> {
        private WordDao wordDao;
        
        private UpdateWordAsyncTask(WordDao wordDao) {
            this.wordDao = wordDao;
        }
        
        @Override
        protected Void doInBackground(Word... words) {
            wordDao.update(words[0]);
            return null;
        }
    }
    
    private static class DeleteWordAsyncTask extends AsyncTask<Word, Void, Void> {
        private WordDao wordDao;
        
        private DeleteWordAsyncTask(WordDao wordDao) {
            this.wordDao = wordDao;
        }
        
        @Override
        protected Void doInBackground(Word... words) {
            wordDao.delete(words[0]);
            return null;
        }
    }
    
    private static class DeleteAllWordsAsyncTask extends AsyncTask<Void, Void, Void> {
        private WordDao wordDao;
        
        private DeleteAllWordsAsyncTask(WordDao wordDao) {
            this.wordDao = wordDao;
        }
        
        @Override
        protected Void doInBackground(Void... voids) {
            wordDao.deleteAllWords();
            return null;
        }
    }
}

File: app/src/main/java/com/example/englishhindi/data/repository/WordRepository.kt
package com.example.englishhindi.data.repository

import androidx.lifecycle.LiveData
import com.example.englishhindi.data.dao.WordDao
import com.example.englishhindi.data.model.Word
import com.example.englishhindi.data.relation.WordWithCategories
import kotlinx.coroutines.flow.Flow
import java.util.Date

class WordRepository(private val wordDao: WordDao) {

    val allWords: Flow<List<Word>> = wordDao.getAllWords()
    val favoriteWords: Flow<List<Word>> = wordDao.getFavoriteWords()
    
    suspend fun insert(word: Word): Long {
        return wordDao.insert(word)
    }
    
    suspend fun update(word: Word) {
        wordDao.update(word)
    }
    
    suspend fun delete(word: Word) {
        wordDao.delete(word)
    }
    
    fun getWordById(id: Long): Flow<Word> {
        return wordDao.getWordById(id)
    }
    
    fun getWordsByDifficulty(difficulty: Int): Flow<List<Word>> {
        return wordDao.getWordsByDifficulty(difficulty)
    }
    
    fun searchWords(query: String): Flow<List<Word>> {
        return wordDao.searchWords("%$query%")
    }
    
    fun getWordWithCategories(wordId: Long): Flow<WordWithCategories> {
        return wordDao.getWordWithCategories(wordId)
    }
    
    fun getAllWordsWithCategories(): Flow<List<WordWithCategories>> {
        return wordDao.getAllWordsWithCategories()
    }
    
    suspend fun toggleFavorite(wordId: Long, isFavorite: Boolean) {
        wordDao.toggleFavorite(wordId, isFavorite)
    }
    
    fun getRecentlyAddedWords(limit: Int): Flow<List<Word>> {
        return wordDao.getRecentlyAddedWords(limit)
    }
    
    fun getWordsByCategoryId(categoryId: Long): Flow<List<Word>> {
        return wordDao.getWordsByCategoryId(categoryId)
    }
    
    fun getWordsAddedBetween(startDate: Date, endDate: Date): Flow<List<Word>> {
        return wordDao.getWordsAddedBetween(startDate, endDate)
    }
    
    fun getWordCount(): Flow<Int> {
        return wordDao.getWordCount()
    }
    
    fun getWordCountByDifficulty(difficulty: Int): Flow<Int> {
        return wordDao.getWordCountByDifficulty(difficulty)
    }
    
    fun getRandomWords(count: Int): Flow<List<Word>> {
        return wordDao.getRandomWords(count)
    }
}